File: py-source/sensors/binary_sensor.py
from ha_mqtt_discoverable import Discoverable
from ha_mqtt_discoverable.sensors import BinarySensorInfo


class BinarySensor(Discoverable[BinarySensorInfo]):
    def off(self):
        """
        Set binary sensor to off
        """
        self.update_state(state=False)

    def on(self):
        """
        Set binary sensor to on
        """
        self.update_state(state=True)

    def update_state(self, state: bool) -> None:
        """
        Update MQTT sensor state

        Args:
            state(bool): What state to set the sensor to
        """
        state_message = self._entity.payload_on if state else self._entity.payload_off
        logger.info(f"Setting {self._entity.name} to {state_message} using {self.state_topic}")
        self._state_helper(state=state_message)

File: src/sensors/binary-sensor/binary-sensor.ts
import { Discoverable } from '../../discoverable';
import { BinarySensorInfo } from './binary-sensor-info';
import { SwitchInfo } from '../switch/switch-info';
import { HaDiscoverableManager } from '../../settings';

/**
 * Implements an MQTT binary sensor for Home Assistant discovery
 */
export class BinarySensor extends Discoverable<BinarySensorInfo | SwitchInfo> {
  static [Symbol.hasInstance](val: unknown): val is BinarySensor {
    if (!val) {
      return false;
    }
    if (val.constructor === BinarySensor) {
      return true;
    }
    return (
      val instanceof Discoverable &&
      'switchOff' in val &&
      'switchOff' in val &&
      'updateState' in val &&
      typeof val['switchOff'] === 'function' &&
      typeof val['switchOff'] === 'function' &&
      typeof val['updateState'] === 'function'
    );
  }

  public readonly entity: BinarySensorInfo;

  constructor(entity: BinarySensorInfo, globalSettings: HaDiscoverableManager) {
    super(globalSettings, entity);
    this.entity = entity;
  }

  /**
   * Set binary sensor to off
   * **/
  switchOff(this: Pick<BinarySensor, 'updateState'>) {
    return this.updateState(false);
  }

  /**
   * Set binary sensor to on
   * **/
  switchOn(this: Pick<BinarySensor, 'updateState'>) {
    return this.updateState(true);
  }

  /** Update MQTT sensor state */
  updateState(state: boolean) {
    return this._state_helper(state ? this.entity.payloadOn : this.entity.payloadOff);
  }
}
